File: tests/python/protocol/test_create_topic.py
#!/usr/bin/env python3
"""Test creating a topic before producing messages"""

import struct
import socket

def encode_string(s):
    """Encode a string as Kafka protocol string (length + data)"""
    if s is None:
        return struct.pack('>h', -1)
    encoded = s.encode('utf-8')
    return struct.pack('>h', len(encoded)) + encoded

def send_create_topics_request(host, port, topic_name, num_partitions=1, replication_factor=1):
    """Send a CreateTopics request"""
    
    # Build create topics request body
    body = b''
    body += struct.pack('>i', 1)  # topic_count
    
    # Topic data
    body += encode_string(topic_name)
    body += struct.pack('>i', num_partitions)  # num_partitions
    body += struct.pack('>h', replication_factor)  # replication_factor
    body += struct.pack('>i', 0)  # replica_assignments count
    body += struct.pack('>i', 0)  # config_entries count
    
    body += struct.pack('>i', 30000)  # timeout_ms
    body += struct.pack('>b', 1)  # validate_only = false
    
    # Request header
    header = b''
    header += struct.pack('>h', 19)  # API key (CreateTopics)
    header += struct.pack('>h', 5)  # API version
    header += struct.pack('>i', 789)  # Correlation ID
    header += encode_string('test-client')  # Client ID
    
    # Full request
    request = header + body
    request_with_length = struct.pack('>i', len(request)) + request
    
    # Send request
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.connect((host, port))
    sock.send(request_with_length)
    
    # Read response
    response_length_data = sock.recv(4)
    response_length = struct.unpack('>i', response_length_data)[0]
    response_data = sock.recv(response_length)
    
    # Parse response header
    correlation_id = struct.unpack('>i', response_data[:4])[0]
    
    # Parse response body
    pos = 4
    throttle_time_ms = struct.unpack('>i', response_data[pos:pos+4])[0]
    pos += 4
    
    topic_count = struct.unpack('>i', response_data[pos:pos+4])[0]
    pos += 4
    
    results = []
    for _ in range(topic_count):
        topic_name_len = struct.unpack('>h', response_data[pos:pos+2])[0]
        pos += 2
        topic_name = response_data[pos:pos+topic_name_len].decode('utf-8')
        pos += topic_name_len
        
        error_code = struct.unpack('>h', response_data[pos:pos+2])[0]
        pos += 2
        
        # Skip error message if present
        error_msg_len = struct.unpack('>h', response_data[pos:pos+2])[0]
        pos += 2
        if error_msg_len > 0:
            pos += error_msg_len
            
        results.append({
            'topic': topic_name,
            'error_code': error_code
        })
    
    sock.close()
    return correlation_id, results

# Test creating a topic
print("Creating topic 'test-topic'...")
corr_id, results = send_create_topics_request('localhost', 9092, 'test-topic', 3, 1)
print(f"Correlation ID: {corr_id}")
for result in results:
    print(f"Topic: {result['topic']}, Error Code: {result['error_code']}")
    if result['error_code'] == 0:
        print("✓ Topic created successfully")
    elif result['error_code'] == 36:  # TOPIC_ALREADY_EXISTS
        print("✓ Topic already exists")
    else:
        print(f"✗ Error creating topic: {result['error_code']}")

File: temp_cleanup/TestCreateTopic.java
import java.net.Socket;
import java.nio.ByteBuffer;
import java.io.InputStream;
import java.io.OutputStream;

public class TestCreateTopic {
    public static void main(String[] args) throws Exception {
        Socket socket = new Socket("localhost", 9092);
        OutputStream out = socket.getOutputStream();
        InputStream in = socket.getInputStream();
        
        // Build CreateTopics request v0
        ByteBuffer request = ByteBuffer.allocate(256);
        request.putShort((short) 19);  // API key (CreateTopics)
        request.putShort((short) 0);   // API version 0
        request.putInt(100);            // Correlation ID
        request.putShort((short) -1);  // Client ID (null)
        
        // CreateTopics body
        request.putInt(1);              // 1 topic to create
        
        // Topic name
        String topicName = "test-topic";
        request.putShort((short) topicName.length());
        request.put(topicName.getBytes());
        
        // Partitions
        request.putInt(3);              // 3 partitions
        
        // Replication factor
        request.putShort((short) 1);    // replication factor 1
        
        // Replica assignments (empty array for auto-assignment)
        request.putInt(0);              // 0 assignments
        
        // Config entries (empty for v0)
        request.putInt(0);              // 0 configs
        
        // Timeout
        request.putInt(30000);          // 30 second timeout
        
        // Send request
        int requestSize = request.position();
        ByteBuffer sizeBuffer = ByteBuffer.allocate(4);
        sizeBuffer.putInt(requestSize);
        out.write(sizeBuffer.array());
        out.write(request.array(), 0, requestSize);
        out.flush();
        
        System.out.println("Sent CreateTopics request for topic: " + topicName);
        
        // Read response size
        byte[] sizeBytes = new byte[4];
        in.read(sizeBytes);
        int size = ByteBuffer.wrap(sizeBytes).getInt();
        System.out.println("Response size: " + size);
        
        // Read response
        byte[] response = new byte[size];
        int totalRead = 0;
        while (totalRead < size) {
            int read = in.read(response, totalRead, size - totalRead);
            if (read < 0) break;
            totalRead += read;
        }
        
        // Parse response
        ByteBuffer respBuffer = ByteBuffer.wrap(response);
        int correlationId = respBuffer.getInt();
        System.out.println("Correlation ID: " + correlationId);
        
        // Topic errors array
        int topicCount = respBuffer.getInt();
        System.out.println("Topic count in response: " + topicCount);
        
        if (topicCount > 0) {
            // Topic name
            short nameLen = respBuffer.getShort();
            byte[] nameBytes = new byte[nameLen];
            respBuffer.get(nameBytes);
            String returnedName = new String(nameBytes);
            
            // Error code
            short errorCode = respBuffer.getShort();
            
            System.out.println("Topic: " + returnedName + ", Error code: " + errorCode);
            
            if (errorCode == 0) {
                System.out.println("SUCCESS: Topic created successfully!");
            } else {
                System.out.println("Failed with error code: " + errorCode);
            }
        }
        
        socket.close();
        
        // Now test Metadata again to see if topic appears
        System.out.println("\nFetching metadata to verify topic exists...");
        socket = new Socket("localhost", 9092);
        out = socket.getOutputStream();
        in = socket.getInputStream();
        
        // Metadata request
        request = ByteBuffer.allocate(14);
        request.putShort((short) 3);  // Metadata
        request.putShort((short) 0);  // v0
        request.putInt(101);           // Correlation ID
        request.putShort((short) -1); // Client ID
        request.putInt(-1);           // All topics
        
        sizeBuffer = ByteBuffer.allocate(4);
        sizeBuffer.putInt(request.position());
        out.write(sizeBuffer.array());
        out.write(request.array(), 0, request.position());
        out.flush();
        
        // Read response
        in.read(sizeBytes);
        size = ByteBuffer.wrap(sizeBytes).getInt();
        response = new byte[size];
        totalRead = 0;
        while (totalRead < size) {
            int read = in.read(response, totalRead, size - totalRead);
            if (read < 0) break;
            totalRead += read;
        }
        
        respBuffer = ByteBuffer.wrap(response);
        correlationId = respBuffer.getInt();
        
        // Skip brokers
        int brokersLen = respBuffer.getInt();
        for (int i = 0; i < brokersLen; i++) {
            respBuffer.getInt(); // node id
            short hostLen = respBuffer.getShort();
            respBuffer.position(respBuffer.position() + hostLen); // skip host
            respBuffer.getInt(); // port
        }
        
        // Topics
        int topicsLen = respBuffer.getInt();
        System.out.println("Topics found: " + topicsLen);
        
        for (int i = 0; i < topicsLen; i++) {
            short errorCode = respBuffer.getShort();
            short nameLen = respBuffer.getShort();
            byte[] nameBytes = new byte[nameLen];
            respBuffer.get(nameBytes);
            String name = new String(nameBytes);
            System.out.println("  - " + name + " (error: " + errorCode + ")");
            
            // Skip partitions
            int partitionsLen = respBuffer.getInt();
            for (int j = 0; j < partitionsLen; j++) {
                respBuffer.getShort(); // error
                respBuffer.getInt(); // partition id
                respBuffer.getInt(); // leader
                
                // Replicas array
                int replicasLen = respBuffer.getInt();
                for (int k = 0; k < replicasLen; k++) {
                    respBuffer.getInt();
                }
                
                // ISR array
                int isrLen = respBuffer.getInt();
                for (int k = 0; k < isrLen; k++) {
                    respBuffer.getInt();
                }
            }
        }
        
        socket.close();
    }
}